[Bookings] Unsaved changes dialog to Filters screen

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This adds an unsaved changes dialog to the Bookings Filters screen. When exiting the screen, if the saved filters have been modified, this new dialog will appear.
The KEEP CHANGES button simply hides the dialog.
DISCARD button discards the new changes and returns to the Booking List screen.

The dialog’s style differs from the Orders Filters dialog, but that’s expected, the Bookings screen is built with Compose, and there are also other visual differences between these screens for the same reason.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
1 Log in with a CIAB store.
2. Go to Bookings.
3. Go to the All tab.
4. Tap the Filters button.
5. Select Booking type.
6. Select a type.
7. Navigate back.
8. Tap X (Close) button.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="360" src="https://github.com/user-attachments/assets/54ea8658-3e3a-4d39-806f-5bae7719a4ba" />

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add unsaved changes dialog to booking filters

Introduces a new `UnsavedChangesDialogState` to manage the visibility and actions of a dialog for unsaved changes within the booking filters screen.

irfano committed:
Show unsaved changes dialog in Booking Filter screen

When there are unsaved changes in the booking filters, display a confirmation dialog to the user before discarding them.

irfano committed:
Show unsaved changes dialog when navigating back from booking filters screen

If there are unsaved changes to the booking filters, show a confirmation dialog before exiting the screen.

irfano committed:
Add unit tests for unsaved filter changes dialog

irfano committed:
Remove redundant filter reset in `onDiscardChanges`

Removes the unnecessary clearing of `newBookingFilters` when discarding changes in the booking filter list.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListViewModel.kt:
```diff
@@ -77,6 +89,31 @@ class BookingFilterListViewModel @Inject constructor(
+                unsavedChangesDialog = current.unsavedChangesDialog.copy(isVisible = false),
+                newBookingFilters = emptySet()
+            )
+        }
```
I removed `newBookingFilters` since it’s redundant (807d1ca2f92a5b6984864215470f120731fd1468). However, the logic for hiding the dialog isn’t unnecessary. If we remove it, the dialog remains visible for a short time after exiting the Filters screen.

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListViewModel.kt:
```diff
@@ -77,6 +89,31 @@ class BookingFilterListViewModel @Inject constructor(
+    }
+
+    private fun hasUnsavedChanges(): Boolean {
+        val initial = _uiState.value.initialBookingFilters ?: BookingFilters()
+        val updated = _uiState.value.updatedBookingFilters
+        return updated != initial
```
```kotlin
return _uiState.value.newBookingFilters.isNotEmpty()
```

This should technically be enough.

irfano replied:
Actually, it’s not the same. If I change an option from A to B, then back from B to A, `newBookingFilters` will not be empty, but in reality, there won’t be any unsaved changes.

AdamGrzybkowski replied:
You're right!

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListUiState.kt:
```diff
@@ -20,10 +20,17 @@ enum class BookingFilterPage {
+    val onDismiss: () -> Unit = {},
+    val onDiscardChanges: () -> Unit = {},
+)
+
```
Can we use `DialogState` like we did on other screens? [example](https://github.com/woocommerce/woocommerce-android/blob/trunk/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/details/BookingDetailsViewState.kt#L19)

One thing is to stay consistent but also to reuse what we already have. With the `DialogState`, the only thing we have to add in Composable component is this `viewState.dialogState?.Render()`.

irfano replied:
Good suggestion! I wasn't aware of `DialogState`. I updated the code: 36cf5ab84eda6afc8bcf1346b66c9ef3352057eb

AdamGrzybkowski reviewed (commented):
Works good! I think it would be worth using the `DialogState` instead, as mentioned in the comment.

irfano committed:
Use generic DialogState for unsaved changes dialog

Replaces the specific `UnsavedChangesDialogState` with the more generic `DialogState` component for handling the "unsaved changes" dialog in the booking filters screen.

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/BookingFilterListUiState.kt:
```diff
@@ -20,17 +21,11 @@ enum class BookingFilterPage {
     val newBookingFilters: Set<BookingsFilterOption> = emptySet(),
     val currentPage: BookingFilterPage = BookingFilterPage.List,
-    val unsavedChangesDialog: UnsavedChangesDialogState = UnsavedChangesDialogState(),
+    val unsavedChangesDialogState: DialogState? = null,
```
np: We can keep a generic `dialogState` property name in case we have other Dialogs to show later.

irfano replied:
Done: a1d98fae2c09f4e2cbb2c89cb89dae431cd474cd

AdamGrzybkowski reviewed (approved):
Thanks for the changes! :shipit:

irfano committed:
Rename `unsavedChangesDialogState` to `dialogState`

Renames the `unsavedChangesDialogState` property to the more generic `dialogState` within the booking filter feature.
